subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/writer/WriteDependencyVerificationFile.java

```diff
@@ -502,57 +518,66 @@ private void exportKeyRingCollection(PublicKeyService publicKeyService, BuildTre
             .filter(WriteDependencyVerificationFile::hasAtLeastOnePublicKey)
             .filter(e -> existingRings.stream().noneMatch(ring -> keyIds(ring).equals(keyIds(e))))
             .collect(Collectors.toList());
-        ImmutableList<PGPPublicKeyRing> allKeyRings = ImmutableList.<PGPPublicKeyRing>builder()
+
+        return ImmutableList.<PGPPublicKeyRing>builder()
             .addAll(existingRings)
             .addAll(keysSeenInVerifier)
             .build();
-        File keyringFile = keyrings.getBinaryKeyringsFile();
-        writeBinaryKeyringFile(keyringFile, allKeyRings);
-        File asciiArmoredFile = keyrings.getAsciiKeyringsFile();
-        writeAsciiArmoredKeyRingFile(asciiArmoredFile, allKeyRings);
-        LOGGER.lifecycle("Exported {} keys to {} and {}", allKeyRings.size(), keyringFile, asciiArmoredFile);
     }
 
-    private void writeAsciiArmoredKeyRingFile(File ascii, ImmutableList<PGPPublicKeyRing> allKeyRings) throws IOException {
+    protected static Collection<PGPPublicKey> gatherPublicKeys(Collection<PGPPublicKeyRing> keyRings) {
+        // Extract all the distinct public keys from the keyrings, and deduplicate them by the keyID
+        Map<Long, PGPPublicKey> allKeys = keyRings.stream()
+            .map(PGPPublicKeyRing::getPublicKeys)
+            .flatMap(keyIter -> {
+                Spliterator<PGPPublicKey> keySpliterator = Spliterators.spliteratorUnknownSize(keyIter, Spliterator.SIZED | Spliterator.IMMUTABLE);
+                return StreamSupport.stream(keySpliterator, false);
+            })
+            .collect(Collectors.toMap(PGPPublicKey::getKeyID, Function.identity(), (a, b) -> {
+                    LOGGER.debug("Found duplicated verification key: " + a.getKeyID());
+                    return a;
+                }
+            ));
+
+        return allKeys.values();
+    }
+
+    private void writeAsciiArmoredKeyRingFile(File ascii, Iterable<PGPPublicKey> allKeys) throws IOException {
         if (ascii.exists()) {
             ascii.delete();
         }
         boolean hasKey = false;
-        for (PGPPublicKeyRing keyRing : allKeyRings) {
+        for (PGPPublicKey key : allKeys) {
```

Comment:
Since `gatherPublicKeys` method uses `Map`, I wonder what is the order of keys here? Do we keep the order or any new write will shuffle keys (bad for checking git diff)?

Reply:
First of all, you are right: Map#values() doesn't make any guarantees about ordering.
What I also realized is that this is not as simple as calling `sort()`... I've seen that different keys can have a different amount of information, depending on the source. This means that between two serializations, even if we sort by the key's ID, we can end up with two different representations, and messing with the diff.

I will need to think about this a little.